### ISSUE TITLE
Implementing an @export directive

### ISSUE THREAD

**Issue by mariusbancila** [graphql-dotnet/graphql-dotnet#3283]
I want to implement an `@export` directive. This would export the value of a field into a variable that can be used in another part of the query (discussed [here](https://blog.logrocket.com/creating-an-export-graphql-directive/) and [here](https://github.com/graphql/graphql-spec/issues/583)). I have read the documentation about implementing a directive, from [here](https://graphql-dotnet.github.io/docs/getting-started/directives/).

Here is an example of how this would work:
```
query($cid : Int!, $ano : Int!, $customerId : Int)
{
  useCompany(no: $cid) 
  {
    associate(filter : {associateNo : {_eq: $ano}})
    {
      items 
      {
        customerNo @export(as: "customerId")
      }
    }
    
    order(first : 5, filter : {customerNo : {_eq : $customerId}})
    {
      items
      {
        orderNo
        orderDate
        customerNo
      }
    }
  }
}
```
Here, I first execute a query that retrieves a customer ID and then this is used in a second query to fetch some orders of that customer.

I defined a directive as follows:
```
   internal class ExportDirective : Directive
   {
      public ExportDirective() : 
         base("export", DirectiveLocation.Field, DirectiveLocation.FragmentSpread, DirectiveLocation.InlineFragment)
      {
         Description = "Export the value of a field into a variable that can be used somewhere else in the query.";
         Arguments = new QueryArguments(
            new QueryArgument<StringGraphType>
            {
               Name = "as",
               Description = "The name of the exported variable"
            });
      }
   }
```

And a visitor as follows:
```
   internal class ExportDirectiveVisitor : BaseSchemaNodeVisitor
   {
      public override void VisitObjectFieldDefinition(FieldType field, IObjectGraphType type, ISchema schema)
      {
         var applied = field.FindAppliedDirective("export");
         if (applied != null)
         {
            var inner = field.Resolver ?? NameFieldResolver.Instance;
            field.Resolver = new FuncFieldResolver<object>(async context =>
            {
               object result = await inner.ResolveAsync(context);

               var arg = applied.FindArgument("as");
               var exportName = arg?.Name;

               if (!string.IsNullOrEmpty(exportName))
               {
                  var value = context.Variables.ValueFor(exportName);
                  if (value is null)
                  {
                     context.Variables.Add(new Variable(exportName) { Value = result });
                  }
               }

               return result;
            });
         }
      }
   }
```
(Ignore the actual details of this implementation. Might not be correct or complete, but I didn't get a chance to test it.)

I registered them with the schema:
```
   public class MySchema : Schema
   {
      public MySchema(IServiceProvider provider)
          : base(provider)
      {
         Directives.Register(new ExportDirective());
         RegisterVisitor(new ExportDirectiveVisitor());
      }
   }
```

The visitor executes when I start the app and the schema is created. But this is not what I want. I want this to be evaluated every time I run a query. What do I have to do for that?

**Comment by Shane32** [comment]
You want field middleware, not a visitor.

Looking over your objective, here's how I would do it:

1. Define a new directive (already done)
2. Specify a serial execution order (easy, and already the default for mutations)
3. Add a validation rule to validate that fields with `@export` directives defined on the query refer to defined variables within the query, and that the field type is a compatible scalar type (or list-scalar type) with the variable type
4. Add field middleware to take the returned value of fields that have an `@export` clause on them and inject the results into the GraphQL variables

Should not really be too hard.  The biggest problem I foresee is that if the query requires a non-null variable, then the graph will need to be defined with a non-null variable, and the variable validation will fail if a variable is not provided.  I'd probably try modifying step 3 (the validation rule) to make it look through the variables, ensure that the specified variable is not listed, and add it to the Inputs object as a variable with a default value (e.g. 0 for Int).  Then it will pass the variable deserialization, which occurs next I believe.  However, adding a value to an Inputs object may not be supported, especially if there were no variables defined, in which case inputs would refer to Inputs.Empty.  Perhaps it is possible to derive from Inputs and/or ensure it is set within a `ConfigureExecutionOptions` call.  Anyway, if all that fails, your users would simply have to set a default value in their query, like defining the variable in your example above to be `$customerId: Int! = 0`.  Again, these changes would only be necessary if the type was needed to be a non-null type.

**Comment by mariusbancila** [comment]
I'm trying your suggestion. But how do I check that a field has the `export` directive applied to it? Because this doesn't work. `FindAppliedDirective` returns `null`.

```
public class ExportFieldMiddlewear : IFieldMiddleware
{
   public async ValueTask<object> ResolveAsync(IResolveFieldContext context, FieldMiddlewareDelegate next)
   {
      FieldType field = context.FieldDefinition;

      var applied = field.FindAppliedDirective("export");
      if (applied != null)
      {
      }

      return await next(context);
   }
}
```

**Comment by mariusbancila** [comment]
I did as you suggested and wrote this:

```
public class ExportFieldMiddlewear : IFieldMiddleware
{
   public async ValueTask<object> ResolveAsync(IResolveFieldContext context, FieldMiddlewareDelegate next)
   {
      FieldType field = context.FieldDefinition;

      var applied = context.FieldAst.Directives?.Items?.Find(d => d.Name == "export");
      if (applied is object)
      {
         var inner = field.Resolver ?? NameFieldResolver.Instance;
         field.Resolver = new FuncFieldResolver<object>(async context =>
         {
            object result = await inner.ResolveAsync(context);

            var arg = applied.Arguments.FirstOrDefault(a => a.Name == "as");
            string exportName = null;
            if(arg?.Value is GraphQLStringValue s)
               exportName = s.Value.ToString();

            if (!string.IsNullOrEmpty(exportName))
            {
               // update variables
            }

            return result;
         });
      }

      return await next(context);
   }
}
```
The trouble with this is that it does not execute the field resolver the first time I execute the query. After subsequent executions it works without any problem.

**Comment by Shane32** [comment]
how are you adding the middleware?

**Comment by Shane32** [comment]
Middleware wraps existing resolver definitions, and multiple midldewares stack on each other.

How is your schema registered?

If you can supply a copy of your `services.AddGraphQL` code, assuming that's how you set it up, that would be useful.

**Comment by Shane32** [comment]
ok i see the issue.  you're overwriting the field resolver within the middleware.  try this:

```cs
public class ExportFieldMiddlewear : IFieldMiddleware
{
   public async ValueTask<object> ResolveAsync(IResolveFieldContext context, FieldMiddlewareDelegate next)
   {
      FieldType field = context.FieldDefinition;

      var applied = context.FieldAst.Directives?.Items?.Find(d => d.Name == "export");
      if (applied is object)
      {
         //var inner = field.Resolver ?? NameFieldResolver.Instance;
         //field.Resolver = new FuncFieldResolver<object>(async context =>
         //{
            //object result = await inner.ResolveAsync(context);
            object result = await next.ResolveAsync(context);

            var arg = applied.Arguments.FirstOrDefault(a => a.Name == "as");
            string exportName = null;
            if(arg?.Value is GraphQLStringValue s)
               exportName = s.Value.ToString();

            if (!string.IsNullOrEmpty(exportName))
            {
               // update variables
            }

            return result;
         //});
      }

      return await next(context);
   }
}
```

**Comment by sungam3r** [comment]
Hi, was busy to chime in. I think `@export` directive should be implemented as core GraphQL.NET concept. I have seen discussions for a long time about this directive in graphql-spec repo. I don't know current status of it. Unfortunatelly I have no time to track activity in spec repo.

**Comment by Shane32** [comment]
Could you simply do something like this:

```cs
bool set = false;
foreach (var variable in context.Variables)
{
    if (variable.Name == exportName)
    {
        variable.Value = result;
        set = true;
    }
}
if (!set)
{
    context.Variables.Add(new Variable(exportName) { Value = result });
}
```

**Comment by mariusbancila** [comment]
That's what I needed! Thank you for all the help.

**Comment by sungam3r** [comment]
I would like to keep it open in scope of future work for federation support in  v8.

**Comment by mariusbancila** [comment]
I'm bringing this up again because something is not working as expected in my implementation. I have the following query:
```
query($customerId : Int = 0)
{
  associate(filter : {associateNo : {_eq: 1}})
  {
    items 
    {
      customerNo @export(as: "customerId")
    }
  }
  
  order(filter : {customerNo : {_eq : $customerId}})
  {
    items
    {
      orderNo
      orderDate
      customerNo
    }
  }
}
```

My expectations are that:
1.  the query to the associates table is executed first
2. the `@export` directive is evaluated and the `customerId` variable resolved
3. the query to the order table is executed 

What happens instead is:
1.  the query to the associates table is executed first
2. the query to the order table is executed 
3. the `@export` directive is evaluated and the `customerId` variable resolved

What I have is the following:

`ExportDirective`
```
   internal class ExportDirective : Directive
   {
      public ExportDirective() : 
         base("export", DirectiveLocation.Field, DirectiveLocation.FragmentSpread, DirectiveLocation.InlineFragment)
      {
         Description = "Export the value of a field into a variable that can be used somewhere else in the query.";
         Arguments = new QueryArguments(
            new QueryArgument<StringGraphType>
            {
               Name = "as",
               Description = "The name of the exported variable"
            });
      }
   }
```

`FieldMiddlewear`:
```
   public class ExportFieldMiddlewear : IFieldMiddleware
   {
      public async ValueTask<object> ResolveAsync(IResolveFieldContext context, FieldMiddlewareDelegate next)
      {
         FieldType field = context.FieldDefinition;

         var applied = context.FieldAst.Directives?.Items?.Find(d => d.Name == "export");
         if (applied is object)
         {
            object result = await next(context);

            var arg = applied.Arguments.FirstOrDefault(a => a.Name == "as");
            string exportName = null;
            if(arg?.Value is GraphQLStringValue s)
               exportName = s.Value.ToString();

            if (!string.IsNullOrEmpty(exportName))
            {
               bool found = false;
               foreach (var variable in context.Variables)
               {
                  if (variable.Name == exportName)
                  {
                     variable.Value = result;
                     variable.IsDefault = false;
                     found = true;
                  }
               }

               if (!found)
               {
                  context.Variables.Add(new Variable(exportName) { Value = result });
               }
            }

            return result;
         }

         return await next(context);
      }
   }
```

`MySchema`:
```
   public class MySchema : Schema
   {
      public MySchema(IServiceProvider provider, ExportFieldMiddlewear middlewear)
          : base(provider)
      {
         Query = provider.GetRequiredService<MyQueryType>();
         Mutation = provider.GetRequiredService<MyMutationType>();

         Directives.Register(new ExportDirective());
         FieldMiddleware.Use(middlewear);
      }
   }
```

And setup with serial execution strategy for both queries and mutations:
```
services.AddGraphQL(builder => builder
    .ConfigureExecutionOptions(options =>
    {
       // ...
    })
    .AddSystemTextJson()
    .AddDataLoader()
    .AddSchema<MySchema>()
    .AddGraphTypes(typeof(MySchema).Assembly)
    .AddExecutionStrategy<SerialExecutionStrategy>(OperationType.Query)
    .AddExecutionStrategy<SerialExecutionStrategy>(OperationType.Mutation));
```

**Comment by Shane32** [comment]
Do you use a data loader to populate the associate items?

**Comment by mariusbancila** [comment]
Yes, I do use the data loader. The order of the field resolving in the `ExportFieldMiddlewear` is:

1. associate (after this an SQL query executes)
2. items (for associate)
3. order (after this an SQL query executes)
4. items (for order)
5. customerNo (for associate)
6. orderNo (for order)
7. orderDate (for order)
8. customerNo (for order)

This is not the order I expected.

**Comment by Shane32** [comment]
A data loader, by design, executes all other waiting nodes before continuing execution.  This is in order to gather as many nodes as possible to pass into the data loader.  I'm not exactly sure of your design, but for example:

1. `associate` executes and runs a query, returning `List<IDataLoaderResult<Associate>>`
2. `items` executes and unwraps the list
3. GraphQL.NET sees that the list items are `IDataLoaderResult<T>` nodes so it stashes them and looks for the next node to execute.
4. `order` is the only other pending node, so it gets executed, returning `List<IDataLoaderResult<Order>>`
5. the `items` field of `order` is now executed, as the only pending non-dataloader node, which unwraps the list
6. GraphQL.NET sees that the list items are `IDataLoaderResult<T>` nodes so it adds them to its list of data loader nodes to be executed
7. GraphQL.NET cannot find any other non-dataloader nodes to execute, so then it looks at its list of pending data loader nodes
8. GraphQL.NET resolves all of the data loader nodes in the order they were added, adding the results back to the pending node list.  The `IDataLoaderResult<T>` implementation returned by the data loader then would likely hit the database to retrieve all of the necessary data for the queued data loader rows.  In your case, perhaps twice - once for the associate items and once for the order items.
9. GraphQL.NET now has a number of nodes waiting to be executed -- namely, the associate item and the order item.  It executes them each in order that they were queued for execution.

This example would tie perfectly to your description of what is happening now.

**Comment by Shane32** [comment]
There are various solutions here, probably, but no clear answer.

You can simply wrap the order resolver in a SimpleDataLoader so that execution is delayed until after data loaders execute.  But this is only effective one level deep.  So if the `@export` is from a nested node that requires 2 data loader executions to reach it, this solution won't work.  Also, you'd have to manually add this to any field that you expect a caller to use `@export` with.  Perhaps you could get creative with field middleware, automatically wrapping field resolvers with a `SimpleDataLoader` if it detects a variable usage that matches an `@export` directive... but this doesn't solve the first limitation.

A more advanced answer might be to write a custom execution strategy that has support for a second level of data loaders.  Perhaps with a `@delay` directive, fields that use exported variables are added like data loader fields to a third list, which executes only after all regular fields and data loaded fields are done.  In this case perhaps your query looks like this:

```gql
query($customerId : Int = 0)
{
  associate(filter : {associateNo : {_eq: 1}})
  {
    items 
    {
      customerNo @export(as: "customerId")
    }
  }
  
  order(filter : {customerNo : {_eq : $customerId}}) @delay
  {
    items
    {
      orderNo
      orderDate
      customerNo
    }
  }
}
```

It is likely you could write some logic to automatically identify consumers of exported variables and add the `@delay` (or equivalent) without requiring the caller to do so.

**Comment by mariusbancila** [comment]
I have solved this by writing a custom serial strategy. What it does is parsing the query AST first to identify the fields that use exported variables. These are put on a second list of nodes to be executed after all the nodes, including those using the data loader, that don't use exported variables are executed. Both queries and mutations seem to work.

This is the `ExecuteNodeTreeAsync` method of the custom strategy:
```
public override async Task ExecuteNodeTreeAsync(ExecutionContext context, ExecutionNode rootNode)
{
   List<string> exportedVariables = new();
   List<GraphQLField> fieldsUsingExportedVariables = new();

   if (context.Document is object)
   {
      var ops = context.Document.Definitions.OfType<GraphQLOperationDefinition>();

      // first pass, find the exported variables
      foreach (var operation in ops)
      {               
         foreach (var selection in operation.SelectionSet.Selections)
         {
            if (selection is GraphQLField fwl)
            {
               ParseGraphQLField(fwl, exportedVariables);
            }
         }
      }

      // second pass, find the fields using the exported variables
      foreach (var operation in ops)
      {
         foreach (var selection in operation.SelectionSet.Selections)
         {
            if (selection is GraphQLField fwl)
            {
               FindFieldsUsingExportedVariables(fwl, exportedVariables, fieldsUsingExportedVariables);
            }
         }
      }
   }

   // Use a stack to track all nodes in the tree that need to be executed
   var nodes = System.Threading.Interlocked.Exchange(ref _reusableNodes, null) ?? new Stack<ExecutionNode>();
   var delayedNodes = System.Threading.Interlocked.Exchange(ref _reusableDelayedNodes, null) ?? new Stack<ExecutionNode>();
   nodes.Push(rootNode);
   var dataLoaderNodes = System.Threading.Interlocked.Exchange(ref _reusableDataLoaderNodes, null) ?? new Queue<ExecutionNode>();
   var addlNodes = System.Threading.Interlocked.Exchange(ref _reusableAddlNodes, null) ?? new Stack<ExecutionNode>();

   try
   {
      // Process each node on the stack one by one
      while (nodes.Count > 0 || dataLoaderNodes.Count > 0)
      {
         while (nodes.Count > 0)
         {
            var node = nodes.Pop();
            await ExecuteNodeAsync(context, node).ConfigureAwait(false);

            // Push any child nodes on top of the stack
            if (node.Result is IDataLoaderResult)
            {
               dataLoaderNodes.Enqueue(node);
            }
            else if (node is IParentExecutionNode parentNode)
            {
               // Add in reverse order so fields are executed in the correct order
               ApplyToChildren(
                  parentNode, 
                  fieldsUsingExportedVariables,
                  (node, state) => state.Push(node), nodes,
                  (node, state) => state.Push(node), delayedNodes,
                  reverse: true);
            }
         }

         while (dataLoaderNodes.Count > 0)
         {
            var node = dataLoaderNodes.Dequeue();
            await CompleteDataLoaderNodeAsync(context, node).ConfigureAwait(false);

            // Push any child nodes on top of the stack
            if (node.Result is IDataLoaderResult)
            {
               dataLoaderNodes.Enqueue(node);
            }
            else if (node is IParentExecutionNode parentNode)
            {
               // Do not reverse the order of the nodes here
               parentNode.ApplyToChildren((node, state) => state.Push(node), addlNodes, reverse: false);
            }
         }

         // Reverse order of queued nodes from data loader nodes so they are executed in the correct order
         while (addlNodes.Count > 0)
         {
            nodes.Push(addlNodes.Pop());
         }
      }
      
      while (delayedNodes.Count > 0 || dataLoaderNodes.Count > 0)
      {
         while (delayedNodes.Count > 0)
         {
            var node = delayedNodes.Pop();
            await ExecuteNodeAsync(context, node).ConfigureAwait(false);

            // Push any child nodes on top of the stack
            if (node.Result is IDataLoaderResult)
            {
               dataLoaderNodes.Enqueue(node);
            }
            else if (node is IParentExecutionNode parentNode)
            {
               // Add in reverse order so fields are executed in the correct order
               parentNode.ApplyToChildren((node, state) => state.Push(node), delayedNodes, reverse: true);
            }
         }

         while (dataLoaderNodes.Count > 0)
         {
            var node = dataLoaderNodes.Dequeue();
            await CompleteDataLoaderNodeAsync(context, node).ConfigureAwait(false);

            // Push any child nodes on top of the stack
            if (node.Result is IDataLoaderResult)
            {
               dataLoaderNodes.Enqueue(node);
            }
            else if (node is IParentExecutionNode parentNode)
            {
               // Do not reverse the order of the nodes here
               parentNode.ApplyToChildren((node, state) => state.Push(node), addlNodes, reverse: false);
            }
         }

         // Reverse order of queued nodes from data loader nodes so they are executed in the correct order
         while (addlNodes.Count > 0)
         {
            delayedNodes.Push(addlNodes.Pop());
         }
      }            
   }
   finally
   {
      nodes.Clear();
      delayedNodes.Clear();
      dataLoaderNodes.Clear();
      addlNodes.Clear();

      System.Threading.Interlocked.CompareExchange(ref _reusableNodes, nodes, null);
      System.Threading.Interlocked.CompareExchange(ref _reusableDelayedNodes, delayedNodes, null);
      System.Threading.Interlocked.CompareExchange(ref _reusableDataLoaderNodes, dataLoaderNodes, null);
      System.Threading.Interlocked.CompareExchange(ref _reusableAddlNodes, addlNodes, null);
   }
}
```

There are three things going on here:

1. first, identify the exported variables and the nodes that use them
2. execute all the nodes that don't use exported variables (and their children)
3. execute all the nodes that use exported variables (and their children)

BTW, this is how I find what variables are exported from a query:
```
private void ParseGraphQLField(GraphQLField field, List<string> exportedVariables)
{
   if (field is null)
      return;

   if (field.Directives is object)
   {
      foreach (GraphQLDirective item in field.Directives.Items)
      {
         if (item.Name == "export" && item.Arguments is object)
         {
            foreach (var arg in item.Arguments)
            {
               if (arg.Value is GraphQLStringValue sv)
                  exportedVariables.Add(sv.Value.ToString());
            }
         }
      }
   }

   if (field.SelectionSet?.Selections is object)
   {
      foreach (var selection in field.SelectionSet?.Selections)
      {
         if (selection is GraphQLField fwl)
         {
            ParseGraphQLField(fwl, exportedVariables);
         }
      }
   }
}
```

And this is how I find what nodes are using these exported variables:
```
private bool RefersExportedVariable(GraphQLValue obj, List<string> exportedVariables)
{
   if (obj is null)
      return false;

   if (obj is GraphQLVariable v)
   {
      if (exportedVariables.Contains(v.Name.StringValue))
         return true;
   }
   else if (obj is GraphQLObjectValue o)
   {
      if (o.Fields is object)
      {
         foreach (var of in o.Fields)
         {
            if(RefersExportedVariable(of.Value, exportedVariables))
               return true;
         }
      }
   }
   else if(obj is GraphQLListValue l)
   {
      foreach(var item in l.Values)
      {
         if (RefersExportedVariable(item, exportedVariables))
            return true;
      }
   }

   return false;
}

private void FindFieldsUsingExportedVariables(GraphQLField field, List<string> exportedVariables, List<GraphQLField> fieldsUsingExportedVariables)
{
   if (field is null)
      return;

   if (field.Arguments is object)
   {
      foreach (GraphQLArgument arg in field.Arguments)
      {
         if(RefersExportedVariable(arg.Value, exportedVariables))
            fieldsUsingExportedVariables.Add(field);
      }
   }

   if (field.SelectionSet?.Selections is object)
   {
      foreach (var selection in field.SelectionSet?.Selections)
      {
         if (selection is GraphQLField fwl)
         {
            FindFieldsUsingExportedVariables(fwl, exportedVariables, fieldsUsingExportedVariables);
         }
      }
   }
}
```

So far it seems to be working!

**Comment by mariusbancila** [comment]
Excellent suggestion, the state can be anything here. Yes, I tried it an it works so I can remove some unnecessary code.

**Comment by Shane32** [comment]
I also don't quite understand why the delay nodes execute before exhausting the main loop.  I had expected your code to execute all the regular nodes first, then all the data loader nodes, then all the delayed nodes.  Something like this:

```cs
while (nodes.Count > 0 || dataLoaderNodes.Count > 0 || delayedNodes.Count > 0) {
  while (nodes.Count > 0 || dataLoaderNodes.Count > 0) {
    while (nodes.Count > 0) {
      // process regular nodes
    }
    while (dataLoaderNodes.Count > 0) {
      // process data loader nodes
    }
  }
  while (delayedNodes.Count > 0) {
    // move delayed nodes to regular queue
    var node = delayedNodes.Pop();
    nodes.Push(node);
  }
}
```

You may be able to considerably reduce the amount of code in your class with that logic, if it works correctly.

**Comment by mariusbancila** [comment]
I think my implementation is equivalent to yours if we only consider one level of nodes. I first execute all the regular nodes, then all the data loader nodes, then all the delayed nodes. But in your approach the delayed nodes can produce more delayed nodes, which is not supported in my approach. I will try your strategy for handling the nodes.

**Comment by Shane32** [comment]
Glad you got it working, and thanks for posting the source for anyone else looking to implement `@export`!

**Comment by mariusbancila** [comment]
The optimized core of the `ExecuteNodeTreeAsync` method, based on your suggestion is as follows:

```
try
{
   // Process each node on the stack one by one
   while (nodes.Count > 0 || dataLoaderNodes.Count > 0 || delayedNodes.Count > 0)
   {
      while (nodes.Count > 0 || dataLoaderNodes.Count > 0)
      {
         while (nodes.Count > 0)
         {
            var node = nodes.Pop();
            await ExecuteNodeAsync(context, node).ConfigureAwait(false);

            // Push any child nodes on top of the stack
            if (node.Result is IDataLoaderResult)
            {
               dataLoaderNodes.Enqueue(node);
            }
            else if (node is IParentExecutionNode parentNode)
            {
               // Add in reverse order so fields are executed in the correct order
               parentNode.ApplyToChildren(
                  (node, state) => (state.fieldsUsingExportedVariables.Contains(node.Field) ? state.delayedNodes : state.nodes)
                      .Push(node),
                  (nodes, delayedNodes, fieldsUsingExportedVariables),
                  reverse: true);
            }
         }

         while (dataLoaderNodes.Count > 0)
         {
            var node = dataLoaderNodes.Dequeue();
            await CompleteDataLoaderNodeAsync(context, node).ConfigureAwait(false);

            // Push any child nodes on top of the stack
            if (node.Result is IDataLoaderResult)
            {
               dataLoaderNodes.Enqueue(node);
            }
            else if (node is IParentExecutionNode parentNode)
            {
               // Do not reverse the order of the nodes here
               parentNode.ApplyToChildren((node, state) => state.Push(node), addlNodes, reverse: false);
            }
         }

         // Reverse order of queued nodes from data loader nodes so they are executed in the correct order
         while (addlNodes.Count > 0)
         {
            nodes.Push(addlNodes.Pop());
         }
      }

      while (delayedNodes.Count > 0)
      {
         // move delayed nodes to regular queue
         var node = delayedNodes.Pop();
         nodes.Push(node);
      }
   }
}
finally
{
   nodes.Clear();
   delayedNodes.Clear();
   dataLoaderNodes.Clear();
   addlNodes.Clear();

   System.Threading.Interlocked.CompareExchange(ref _reusableNodes, nodes, null);
   System.Threading.Interlocked.CompareExchange(ref _reusableDelayedNodes, delayedNodes, null);
   System.Threading.Interlocked.CompareExchange(ref _reusableDataLoaderNodes, dataLoaderNodes, null);
   System.Threading.Interlocked.CompareExchange(ref _reusableAddlNodes, addlNodes, null);
}
```

Yes, maybe this will help others that need to do the same. 🙋

**Comment by Shane32** [comment]
Some final review notes if this ever happens to get merged in as an official feature:

1. I am guessing that the execution order of the delayed nodes is incorrect, the fix probably being to change `delayedNodes` to a `Queue` instead of a `Stack`.  (Some slight change to the `ApplyToChildren` call is necessary.)  A test should be added to confirm the correct execution order.

2. The code within `while (addlNodes.Count > 0)` should also check if the node uses an exported variable and add it to either `nodes` or `delayedNodes` appropriately.

3. `delayedNodes` could be a reusable instance (like the other lists) for an extremely small performance benefit.

4. `delayedNodes` could be a `HashSet`, but assuming the list is typically very small, a `List` would probably perform better and have less memory requirements

5. `is object` could probably be changed to `!= null` for consistency in the codebase, although they probably execute identically

6. `if (context.Document is object)` should not be necessary as it is not marked as nullable (confirm)
